### PR TITLE
refactor!: conform CRUD result types to specification

### PR DESCRIPTION
**Pull request by mbroadst**
Our CRUD result types have been out-of-sync with the official drivers CRUD specification for some time. This patch brings them in line, and updates our tests accordingly

NODE-2936

### CODE REVIEW

File: src/bulk/common.ts [hunk 1]
```diff
@@ -161,6 +161,8 @@ export class Batch {
 export class BulkWriteResult {
   result: BulkResult;
 
+  /** Indicates whether this write result was acknowledged. If not, then all other members of this result will be undefined */
+  // acknowledged: Boolean;
```
Comment (emadum) [inline src/bulk/common.ts:165]: Is this supposed to be commented out?

**Review by emadum — APPROVED** [review]
Very nice! 👍

File: src/bulk/common.ts (hunk 1 above)
Comment (mbroadst) [reply src/bulk/common.ts:165]: yeah this is deferred to forthcoming bulk typing work

File: src/bulk/common.ts [hunk 2]
```diff
@@ -161,6 +161,8 @@ export class Batch {
 export class BulkWriteResult {
   result: BulkResult;
```
Comment (nbbeeken) [inline src/bulk/common.ts:162]: This still has the undefined ids issue, I think they're not being added in correctly?
I reran the snippet here:
https://jira.mongodb.org/browse/NODE-2923

File: src/operations/insert.ts [hunk 3]
```diff
@@ -67,63 +69,30 @@ export class InsertOneOperation extends CommandOperation<InsertOneOptions, Inser
...
-  docs = Array.isArray(docs) ? docs : [docs];
-
-  // Final options for retryable writes
-  let finalOptions = Object.assign({}, options);
-  finalOptions = applyRetryableWrites(finalOptions, coll.s.db);
-
-  // If keep going set unordered
-  if (finalOptions.keepGoing === true) finalOptions.ordered = false;
+    // If keep going set unordered
+    // if (finalOptions.keepGoing === true) finalOptions.ordered = false;
```
Comment (nbbeeken) [inline src/operations/insert.ts:77]: Should this still be here?

File: src/operations/insert.ts [hunk 4]
```diff
@@ -67,63 +69,30 @@ export class InsertOneOperation extends CommandOperation<InsertOneOptions, Inser
...
     const options = { ...this.options, ...this.bsonOptions };
 
-    if (Array.isArray(doc)) {
-      return callback(
-        MongoError.create({ message: 'doc parameter must be an object', driver: true })
-      );
-    }
+    // Final options for retryable writes
+    // let finalOptions = Object.assign({}, options);
+    // finalOptions = applyRetryableWrites(finalOptions, coll.s.db);
```
Comment (nbbeeken) [inline src/operations/insert.ts:74]: Do retryable writes get handled elsewhere?

**Review by nbbeeken — COMMENTED** [review]
some Qs

File: src/bulk/common.ts (hunk 2 above)
Comment (mbroadst) [reply src/bulk/common.ts:162]: That was fixed in this PR: https://github.com/mongodb/node-mongodb-native/pull/2653

File: src/operations/insert.ts (hunk 4 above)
Comment (mbroadst) [reply src/operations/insert.ts:74]: Yeah, they are handled in [executeOperation](https://github.com/mongodb/node-mongodb-native/blob/master/src/operations/execute_operation.ts#L209). We were actually out of [spec compliance](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#retrywrites) by allowing it to be set at the operation level.

File: src/operations/insert.ts (hunk 3 above)
Comment (mbroadst) [reply src/operations/insert.ts:77]: whoops, nope! Removed

File: src/operations/insert.ts (hunk 4 above)
Comment (nbbeeken) [reply src/operations/insert.ts]: Ah yep, looks like it.

**Review by nbbeeken — APPROVED** [review]

**Pull request merged**